gpflow/models/training_mixins.py

```diff
@@ -0,0 +1,131 @@
+        """
+        return self._training_loss()
+
+    def training_loss_closure(self, *, jit=True) -> Callable[[], tf.Tensor]:
+        """
+        Convenience method. Returns a closure which itself returns the training loss. This closure
+        can be passed to the minimize methods on :class:`gpflow.optimizers.Scipy` and subclasses of
+        `tf.optimizers.Optimizer`.
+
+        :param jit: If `True` (default), compile the training loss function in a TensorFlow graph
+            by wrapping it in tf.function()
+        """
+        if jit:
+            return tf.function(self.training_loss)
+        return self.training_loss
+
+
+class ExternalDataTrainingLossMixin:
+    """
+    Mixin utility for training loss methods for models that do **not** own their own data.
+    It provides
+
+      - a uniform API for the training loss :meth:`training_loss`
+      - a convenience method :meth:`training_loss_closure` for constructing the closure expected by
+        various optimizers, namely :class:`gpflow.optimizers.Scipy` and subclasses of
+        `tf.optimizers.Optimizer`.
+
+    See :class:`InternalDataTrainingLossMixin` for an equivalent mixin for models that **do** own
+    their own data.
+    """
+
+    def training_loss(self, data: Data) -> tf.Tensor:
+        """
+        Returns the training loss for this model.
+        
+        :param data: the data to be used for computing the model objective.
+        """
+        return self._training_loss(data)
+
+    def training_loss_closure(
+        self, data: Union[Data, DatasetOwnedIterator], *, jit=True,
+    ) -> Callable[[], tf.Tensor]:
+        """
+        Returns a closure that computes the training loss, which by default is
+        wrapped in tf.function(). This can be disabled by passing `jit=False`.
+        
+        :param data: the data to be used by the closure for computing the model
+            objective. Can be the full dataset or an iterator, e.g.
+            `iter(dataset.batch(batch_size))`, where dataset is an instance of
+            tf.data.Dataset.
+        :param jit: if True, wrap training loss in tf.function()
+        """
+        training_loss = self.training_loss
+
+        if isinstance(data, DatasetOwnedIterator):
+            if jit:
+                input_signature = [data.element_spec]
+                training_loss = tf.function(training_loss, input_signature=input_signature)
+
+            def closure():
+                batch = next(data)
+                return training_loss(batch)
+
+        else:
+
+            def closure():
+                return training_loss(data)
+
+            if jit:
+                closure = tf.function(closure)
+
+        return closure
+
+
+class MCMCTrainingLossMixin(InternalDataTrainingLossMixin):
+    def training_loss(self) -> tf.Tensor:
+        """
+        Training loss for gradient-based relaxation of MCMC models.
+        """
+        return -self.log_posterior_density()
```

Comment:
why are we using log posterior density instead of `_training_loss`? Stuff like this is confusing

Reply:
we're sayin gon this model: x is an MCMCTrainingLossMixin, but the model `has a _training_loss`, but we ignore this, and define a new one, that is actually the same!

Reply:
if we stick with `_training_loss` lets delete this class.

Reply:
ok, done